Extinguish Torches

Tatrabbit opened the pull request:
Resolving issue 895.

I was a little confused as to whether to use CustomData or the mCharge of a CellRef. I interpreted your (Zini's) last post in the Development topic to mean "Use a CustomData, as the time remaining is a float, not an int". It was a little ambiguous, I hope I understood correctly.

Tatrabbit committed:
Equipped torches and lights run out of fuel

Tatrabbit committed:
Lights extinguish underwater

jeaye commented on apps/openmw/mwmechanics/actors.cpp:
```diff
@@ -196,6 +199,49 @@
+                float timeRemaining = heldIter->getClass().getRemainingUsageTime(*heldIter);
+
+                // -1 is infinite light source. Other negative values are treated as 0.
+                if(timeRemaining != -1.0f)
```
Time calculations with floats, in general, are not precise. With that said, when using an equality operator, even with the right hand side being constexpr, it is _not safe_ to assume `timeRemaining` will be exactly `-1.0f` when you expect it to be. I'd suggest factoring in `std::numeric_limits<float>::epsilon()` at least (and perhaps turning it into a helper function that others can use).

Tatrabbit replied:
I will factor that in if you think it would be beneficial. The reason it's done as a direct equality comparison is because the only time it should ever return false (duration is -1) is when duration has never been altered in the first place. The time remaining would be read as -1 from the ESM (possibly save file later), and the code here won't change that value, in that case. -1 here would be a special value or more of a magic number, not really an important time. Is there something I'm missing?

Tatrabbit replied:
Actually - It just occurred to me. I was messing about with a -1 duration because that's how the file format is, and because I was originally going for the mCharge field of a CellRef. As long as I'm now using CustomData, it would be a better solution to have a bool value for whether it's an infinite light, held in CustomData. That would solve the problem, and make it a lot more clear what's going on.

jeaye replied:
Using a bool instead sounds good to me. Let's make sure it's initialized in the ctor, too.

zinnschlag replied:
Regarding the -1: The issue with floating point numbers and equality comparison is a myth. You can and most likely will get into precision/rounding issues, if you are using the result of calculations or numeric literals that can not be represented exactly by the floating number types involved. In all other cases using == is perfectly safe.

jeaye replied:
> > -                if(timeRemaining != -1.0f)
> 
>    Regarding the -1: The issue with floating point numbers and equality
>    comparison is a myth. You can and most likely will get into
>    precision/rounding issues, if you are using the result of calculations
>    or numeric literals that can not be represented exactly by the floating
>    number types involved. In all other cases using == is perfectly safe.

That is most definitely incorrect. To assure you, please observe this
example I've written http://ideone.com/8fTBlO -- notice that we're
approaching -1.0f using time decrements (we're using time decrements
here in MW as well), but we never quite hit it and thus loop forever.

-0.999981
-1.00098
-1.00198

That's the transition beyond our escape condition, which is most
definitely not safe. I would agree with rainChu here that a bool or
other form of flag should be used. To clarify, _the only time_ the float
will _reliably_ be -1.0f is if it's explicitly set to it. If this is the
case, we are using a float as both float and a bool; we should just add
a bool.

Tatrabbit replied:
Yes, jeaye, you are correct in your example. However, in this case we don't approach -1 at all. We start there and remain there, or else we start at, say, 150, and decrease, and never hit it. If it starts as positive, and decreases below 0, the light is removed. If it starts at -1, it never decreases and is never removed.

jeaye commented on apps/openmw/mwmechanics/actors.cpp:
```diff
@@ -196,6 +199,49 @@
+                heldIter->getRefData().setCount(0); // remove it
+
+                // ...But, only the player makes a sound.
+                if(isPlayer)
```
Why does only the player's torch play a sound? Strictly because vanilla?

Tatrabbit replied:
Yes, strictly because vanilla. It plays like a GUI sound effect, ignoring environment effects and 3D position, when the player's torch is lost to water, but not when it's lost to duration. NPC torches make no sound when they are lost through either method.

jeaye replied:
Ok, sounds good. We should add an option for a nicer 3D effect in the future, but that's out of the scope of this change.

jeaye commented on apps/openmw/mwworld/class.cpp:
```diff
@@ -132,6 +132,16 @@
+
+    float Class::getRemainingUsageTime (const Ptr& ptr) const
+    {
+        throw std::runtime_error ("class does not support time-based uses.");
```
I've noticed none of the other error messages end in punctuation. Perhaps remove yours simply for consistency?

Tatrabbit replied:
Thank you, I had not spotted that. I'll fix these tomorrow after breakfast (it's bedtime now.)

jeaye commented on apps/openmw/mwclass/light.cpp:
```diff
@@ -16,12 +16,27 @@
+    struct CustomData : public MWWorld::CustomData
+    {
+        float mTime;
+        ///< Time remaining
```
A default ctor should be provided that initializes `mTime`, lest it be left as garbage if someone unknownly default-constructs it. Furthermore, to reduce code duplication while creating instances of `CustomData`, provide a ctor overload that takes a float and initializes `mTime`?

Tatrabbit replied:
I had copied this struct closely from the one in mwclass/npc.cpp, which didn't have overridden default ctor or an overloaded ctor. I assumed I didn't need them, but it wouldn't hurt anything to add them here. I'll do it tomorrow.

zinnschlag replied:
To wrap this up:

Npc CustomData does not need a default constructor, because all its members will be properly initialised by the implicitly generated default constructor. You can add a default constructor here (technically it is correct), though I would not have rejected the pull request because of it being missing.

Using -1 for an infinite duration is okay. Its a safe out-of-band value.

No further changes to the CustomData struct are required.

Then we need the punctuation fixes for the error messages and then we should be good to go.

jeaye replied:
>    CustomData does not need a default constructor, because all its
>    members will be properly initialised by the implicitly generated
>    default constructor. You can add a default constructor here
>    (technically it is correct), though I would not have rejected the pull
>    request because of it being missing.

This is incorrect. As an example, see here http://ideone.com/BEUzqs
To guarantee the float member is initialized properly, you _need_ to
provide a default ctor and initialize it to 0.0f.

Tatrabbit replied:
While your example is perfectly correct, jeaye, he was referring to the CustomData class belonging to an NPC, which has only member variables which were classes or structs with a defined default ctor of their own. The NPC custom data class would therefore have fully initialized members, with the compiler-supplied default ctor.

jeaye replied:
>    While your example is perfectly correct, jeaye, he was referring to the
>    CustomData class belonging to an NPC, which has only member variables
>    which were classes or structs with a defined default ctor of their own.
>    The NPC custom data class would therefore have fully initialized
>    members, with the compiler-supplied default ctor.

The NPC CustomData is not in question here; it should be improved to
properly initialize its members, but that's outside the scope of this
PR. What _is_ within the scope of this PR is the new CustomData struct
declared here. Since it is new code (showing up as green), it should be
written such that it doesn't introduce garbage crash bugs when used.
Otherwise it's not a safe and portable data structure.

jeaye replied:
To expand my point, feel free to consult any online resource, such as
http://stackoverflow.com/questions/1069621/are-members-of-a-c-struct-initialized-to-0-by-default

The fact that you're using it as "only a member variable" by no means
guarantees the float will be zeroed; the compiler is giving you a
handout and you're relying on undefined behavior.

Tatrabbit replied:
It doesn't matter either way - I've already added the ctor. NPC CustomData _does_ properly initialize its members. But, I personally don't think this CustomData absolutely needs a defined ctor. Having one is better than not, but it's both declared and defined within a cpp file, inside an anonymous namespace, and is intended to be constructed solely by a single wrapper function which initializes the variables. I suppose if I really wanted to ensure that I could make a default constructor that doesn't initialize variables (slight optimization, since we know that the wrapper does it), make it private, friend the class to MWClass::Light's wrapper function, but I believe that's overkill for what's supposed to be a simple struct.

Tatrabbit replied:
> The fact that you're using it as "only a member variable" by no means
> guarantees the float will be zeroed; the compiler is giving you a
> handout and you're relying on undefined behavior.

I don't expect it to be zeroed at all. It's constructed only in Light::ensureCustomData, and the member variables within are set using the MWWorld::Ptr of the torch in question.

It doesn't matter either way, anyways. I've added the ctor, as I said.

Tatrabbit committed:
Added a constructor to Light CustomData

Tatrabbit committed:
Fixed punctuation consistency

Tatrabbit commented:
I've added a ctor (Thanks jeaye, you're quite right about it being better practice.), and I've fixed the punctuation. Unless I missed something, I'm once again ready for the pull.
